src/util/CircularBuffer.h

```diff
@@ -0,0 +1,75 @@
+// Copyright 2017 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include <vector>
+
+namespace stellar
+{
+
+using std::vector;
+using std::size_t;
+
+template <typename T> class CircularBuffer
+{
+  public:
+    explicit CircularBuffer(const size_t size) : buffer(size), count(0), start(0)
+    {
+    }
+
+    void
+    push(T elem)
+    {
+        size_t ix = postIncrementLastIndex();
+        buffer[ix] = elem;
+    }
+
+    T&
+    pop()
+    {
+        if (empty())
```

Comment:
I think we can safely expect that nobody will call pop() when empty() is true
Please remove that condition and make calling pop() when empty() a undefined behavior, so this behavior will be consistent with std:: containers.